indico/web/client/js/react/components/Carousel.jsx

```diff
@@ -0,0 +1,193 @@
+            });
+        });
+    }
+
+    componentWillUnmount() {
+        this.divRef.current.style.transition = null;
+    }
+
+    render() {
+        const {disabled} = this.props;
+        return (
+            <div styleName="progress-indicator" className={disabled ? 'disabled' : ''}>
+                <div styleName="progress-bar" ref={this.divRef} />
+            </div>
+        );
+    }
+}
+
+/**
+ * The little dots that can be used to switch panes.
+ */
+function PaneSwitcher({numPanes, onClick, activePane}) {
+    return (
+        <ul styleName="pane-switcher">
+            {_.times(numPanes, n => (
+                <li key={n}
+                    className={activePane === n ? 'active' : ''}
+                    styleName="pane-switcher-item"
+                    onClick={() => onClick(n)} />
+            ))}
+        </ul>
+    );
+}
+
+PaneSwitcher.propTypes = {
+    /** The number of "dots" to show */
+    numPanes: PropTypes.number.isRequired,
+    /** onClick handler (pane index passed as parameter) */
+    onClick: PropTypes.func.isRequired,
+    /** the currently active pane */
+    activePane: PropTypes.number.isRequired
+};
+
+/**
+ * A purely CSS-based carousel widget, which switches between a given number of panes.
+ */
+export default class Carousel extends React.PureComponent {
+    static propTypes = {
+        /** an array of {key, content, delay?} objects which will define each pane */
+        panes: PropTypes.arrayOf(PropTypes.shape({
+            key: PropTypes.string.isRequired,
+            content: PropTypes.node.isRequired,
+            delay: PropTypes.number
+        })).isRequired,
+        /** the delay to wait at each pane, by default */
+        delay: PropTypes.number
+    };
+
+    static defaultProps = {
+        delay: 10
+    };
+
+    state = {
+        /** id of the active pane */
+        activePane: 0,
+        /** this will be set to 'true' if the user has taken control */
+        userOverridden: false
+    };
+
+    componentDidMount() {
+        const {delay: globalDelay, panes: [{delay: paneDelay}]} = this.props;
+        this.timeout = setTimeout(this.switchPane, (paneDelay || globalDelay) * 1000);
+    }
+
+    componentWillUnmount() {
+        // cancel timeout for next transition
+        window.clearTimeout(this.timeout);
+    }
+
+    get numPanes() {
+        const {panes} = this.props;
+        return panes.length;
+    }
+
+    switchPane = () => {
+        let {activePane} = this.state;
+        activePane = (activePane + 1) % this.numPanes;
+        const {panes, delay: globalDelay} = this.props;
+        const {delay: paneDelay} = panes[activePane];
+
+        this.setState({
+            activePane,
+        });
+
+        // Schedule next switch
+        this.timeout = setTimeout(this.switchPane, (paneDelay || globalDelay) * 1000);
+    };
+
+    forcePaneSwitch = (newPane) => {
+        // cancel automatic switching (forever!)
+        window.clearTimeout(this.timeout);
```

Comment:
on unmount we'll call it again. set to null and conditionally call it there?

Reply:
Yeah, that's an option. Even if calling it with `null` as a parameter doesn't break anything.